python/ray/autoscaler/autoscaler.py

```diff
@@ -491,6 +590,76 @@ def with_head_node_ip(cmds):
         out.append("export RAY_HEAD_IP={}; {}".format(head_ip, cmd))
     return out
 
+def with_docker_exec(cmds, container_name=DEFAULT_CONTAINER_NAME, env_vars=None):
+    env_str = ""
+    if env_vars:
+        env_str = " ".join(
+            ["-e {env}=${env}".format(env=env) for env in env_vars])
+    return ["docker exec {} {} /bin/sh -c '{}' ".format(
+        env_str, container_name, cmd) for cmd in cmds]
+
+def try_command(cmd):
+    return "sudo yum {0} || sudo apt-get {0}".format(cmd)
+
+def docker_install_cmds():
+    return [try_command("update"),
+    "sudo DEBIAN_FRONTEND=noninteractive apt-get -y -o Dpkg::Options::='--force-confdef' -o Dpkg::Options::='--force-confold' upgrade || true",
+    "sudo yum install -y docker-ce || sudo apt-get install -y docker.io"]
+
+def docker_start_cmds(user, image, mount, ctnr_name=DEFAULT_CONTAINER_NAME):
+    cmds = []
+    cmds.append("sudo kill -SIGUSR1 $(pidof dockerd) || true")
+    cmds.append("sudo service docker start")
+    cmds.append("sudo usermod -a -G docker {}".format(user))
+    cmds.append("docker rm -f {} || true".format(ctnr_name))
+    cmds.append("docker pull {}".format(image))
+
+    # create flags
+    # ports for the redis, object manager, and tune client
+    port_flags = " ".join(["-p {port}:{port}".format(port=port)
+                              for port in ["6379", "8076", "4321"]])
+    mount_flags = " ".join(["-v {src}:{dest}".format(src=k, dest=v)
+                               for k, v in mount.items()])
+
+    # click ........
+    env_vars = {"LC_ALL": "C.UTF-8", "LANG": "C.UTF-8"}
+    env_flags = " ".join(["-e {name}={val}".format(name=k, val=v)
+                             for k, v in env_vars.items()])
+    # docker run command
+    cmds.append("docker run --name {} -d -it {} {} {} --net=host {} bash".format(
+        ctnr_name, port_flags, mount_flags, env_flags, image))
+    cmds.append("docker exec {} apt-get -y update".format(ctnr_name))
+    cmds.append("docker exec {} apt-get -y upgrade".format(ctnr_name))
+    cmds.append("docker exec {} apt-get install -y git wget cmake psmisc".format(ctnr_name))
+    return cmds
+
+
+def ray_install_cmds():
+    """These commands assume pip is installed."""
+    return ["pip install ray"]
```

Comment:
it would make sense to patch this after merge so that there's a more update to date version of ray being installed